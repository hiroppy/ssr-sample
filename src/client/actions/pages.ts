import { Action } from 'redux';

export type Actions =
  | ResetPagesStatus
  | LoadAppProcess
  | LoadAppProcessSuccess
  | LoadAppProcessFailure
  | LoadTopPage
  | LoadTopPageSuccess
  | LoadTopPageFailure
  | LoadOrgsPage
  | LoadOrgsPageSuccess
  | LoadOrgsPageFailure
  | StopSaga;

export interface ResetPagesStatus extends Action<'RESET_PAGES_STATUS'> {}

export interface LoadAppProcess extends Action<'LOAD_APP_PROCESS'> {}

export const loadAppProcess = (): LoadAppProcess => ({
  type: 'LOAD_APP_PROCESS'
});

export interface LoadAppProcessSuccess extends Action<'LOAD_APP_PROCESS_SUCCESS'> {}

export const loadAppProcessSuccess = (): LoadAppProcessSuccess => ({
  type: 'LOAD_APP_PROCESS_SUCCESS'
});

export interface LoadAppProcessFailure extends Action<'LOAD_APP_PROCESS_FAILURE'> {
  err: Error;
}

export const loadAppProcessFailure = (err: Error): LoadAppProcessFailure => ({
  type: 'LOAD_APP_PROCESS_FAILURE',
  err
});

export const resetPageStatus = (): ResetPagesStatus => ({
  type: 'RESET_PAGES_STATUS'
});

export interface LoadTopPage extends Action<'LOAD_TOP_PAGE'> {}

export const loadTopPage = (): LoadTopPage => ({
  type: 'LOAD_TOP_PAGE'
});

export interface LoadTopPageSuccess extends Action<'LOAD_TOP_PAGE_SUCCESS'> {}

export const loadTopPageSuccess = (): LoadTopPageSuccess => ({
  type: 'LOAD_TOP_PAGE_SUCCESS'
});

export interface LoadTopPageFailure extends Action<'LOAD_TOP_PAGE_FAILURE'> {
  payload: {
    err: Error;
  };
}

export const loadTopPageFailure = (err: Error): LoadTopPageFailure => ({
  type: 'LOAD_TOP_PAGE_FAILURE',
  payload: {
    err
  }
});

export interface LoadOrgsPage extends Action<'LOAD_ORGS_PAGE'> {
  payload: {
    org: string;
  };
}

export const loadOrgsPage = (org: LoadOrgsPage['payload']['org']): LoadOrgsPage => ({
  type: 'LOAD_ORGS_PAGE',
  payload: {
    org
  }
});

export interface LoadOrgsPageSuccess extends Action<'LOAD_ORGS_PAGE_SUCCESS'> {}

export const loadOrgsPageSuccess = (): LoadOrgsPageSuccess => ({
  type: 'LOAD_ORGS_PAGE_SUCCESS'
});

export interface LoadOrgsPageFailure extends Action<'LOAD_ORGS_PAGE_FAILURE'> {
  payload: {
    err: Error;
  };
}

export const loadOrgsPageFailure = (err: Error): LoadOrgsPageFailure => ({
  type: 'LOAD_ORGS_PAGE_FAILURE',
  payload: {
    err
  }
});

export interface StopSaga extends Action<'STOP_SAGA'> {}

export const stopSaga = (): StopSaga => ({
  type: 'STOP_SAGA'
});
