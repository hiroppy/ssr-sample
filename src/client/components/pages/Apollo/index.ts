import { Apollo } from './Apollo';

export default Apollo;
