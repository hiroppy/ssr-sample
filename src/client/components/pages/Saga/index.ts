import { Saga } from './Saga';

export default Saga;
