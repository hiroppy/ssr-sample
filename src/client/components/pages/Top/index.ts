import { Top } from './Top';

export default Top;
