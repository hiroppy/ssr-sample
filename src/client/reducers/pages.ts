import { Actions } from '../actions/pages';

export type State = {
  isLoadingCompletion: boolean;
  error: Error | null;
  baseUrl: string;
  env: typeof process.env.NODE_ENV;
};

export const initialState: State = {
  isLoadingCompletion: false,
  error: null,
  baseUrl: '',
  env: 'development'
};

export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case 'SET_ENV':
      return { ...state, env: action.payload.env };
    case 'RESET_PAGES_STATUS':
      return initialState;
    case 'SET_BASE_URL':
      return { ...state, baseUrl: action.payload.baseUrl };
    case 'LOAD_TOP_PAGE':
    case 'LOAD_SAGA_PAGE':
    case 'LOAD_APOLLO_PAGE':
      return { ...state, isLoadingCompletion: false };
    case 'LOAD_TOP_PAGE_SUCCESS':
    case 'LOAD_SAGA_PAGE_SUCCESS':
    case 'LOAD_APOLLO_PAGE_SUCCESS':
      return { ...state, isLoadingCompletion: true };
    case 'LOAD_TOP_PAGE_FAILURE':
    case 'LOAD_SAGA_PAGE_FAILURE':
    case 'LOAD_APOLLO_PAGE_FAILURE':
      return { ...state, isLoadingCompletion: true, error: action.payload.err };
    default:
      return state;
  }
};
