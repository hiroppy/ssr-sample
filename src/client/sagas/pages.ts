import { END } from 'redux-saga';
import { put, call, take, select, takeLatest } from 'redux-saga/effects';
import {
  loadAppProcessSuccess,
  loadAppProcessFailure,
  loadTopPageSuccess,
  loadTopPageFailure,
  LoadOrgsPage,
  loadOrgsPageSuccess,
  loadOrgsPageFailure,
  resetPageStatus
} from '../actions/pages';
import { setUserName } from '../actions/users';
import { fetchRepos, FetchReposSuccess, FetchReposFailure, resetOrgs } from '../actions/orgs';
import { getOrgs, getUsers } from './selectors';
import { State } from '../reducers';

// don't call `stopSaga`
function* appProcess() {
  try {
    // e.g. write common processing to be performed on all pages
    const users: State['users'] = yield select<State>(getUsers);

    if (users.name === '') {
      yield put(setUserName('hiroppy'));
    }
    yield put(loadAppProcessSuccess());
  } catch (err) {
    yield put(loadAppProcessFailure(err));
    if (!process.env.IS_BROWSER) yield call(stopSaga);
  }
}

function* loadTopPage() {
  try {
    yield put(loadTopPageSuccess());
  } catch (err) {
    yield put(loadTopPageFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) yield call(stopSaga);
  }
}

function* loadOrgsPage(action: LoadOrgsPage) {
  try {
    const { org } = action.payload;
    const orgs: State['orgs'] = yield select(getOrgs);

    if (org !== orgs.name) {
      yield put(resetOrgs());
      yield put(fetchRepos(org));
      const res: FetchReposSuccess | FetchReposFailure = yield take([
        'FETCH_REPOS_SUCCESS',
        'FETCH_REPOS_FAILURE'
      ]);

      if (res.type === 'FETCH_REPOS_FAILURE') throw res.payload;
    }

    yield put(loadOrgsPageSuccess());
  } catch (err) {
    yield put(loadOrgsPageFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) yield call(stopSaga);
  }
}

function* stopSaga() {
  yield put(END);
}

function* changeLocation() {
  yield put(resetPageStatus());
}

export function* pagesProcess() {
  yield takeLatest('LOAD_APP_PROCESS', appProcess);
  yield takeLatest('LOAD_TOP_PAGE', loadTopPage);
  yield takeLatest('LOAD_ORGS_PAGE', loadOrgsPage);
  yield takeLatest('STOP_SAGA', stopSaga);
  yield takeLatest('@@router/LOCATION_CHANGE', changeLocation);
}
