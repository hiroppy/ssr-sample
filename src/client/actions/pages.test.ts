import * as actions from './pages';

test('should create an action to set a base-url', () => {
  expect(actions.setBaseUrl('url')).toMatchSnapshot();
});

test('should create an action to reset state', () => {
  expect(actions.resetPageStatus()).toMatchSnapshot();
});

test('should create an action to load app process', () => {
  expect(actions.loadAppProcess()).toMatchSnapshot();
});

test('should create a success action to load app process', () => {
  expect(actions.loadAppProcessSuccess()).toMatchSnapshot();
});

test('should create a failure action to load app process', () => {
  expect(actions.loadAppProcessFailure(new Error())).toMatchSnapshot();
});

test('should create an action to load Top page', () => {
  expect(actions.loadTopPage()).toMatchSnapshot();
});

test('should create a success action to load Top page', () => {
  expect(actions.loadTopPageSuccess()).toMatchSnapshot();
});

test('should create a failure action to load Top page', () => {
  expect(actions.loadTopPageFailure(new Error())).toMatchSnapshot();
});

test('should create an action to load Saga page', () => {
  expect(actions.loadSagaPage('2')).toMatchSnapshot();
});

test('should create a success action to load Saga page', () => {
  expect(actions.loadSagaPageSuccess()).toMatchSnapshot();
});

test('should create a failure action to load Saga page', () => {
  expect(actions.loadSagaPageFailure(new Error())).toMatchSnapshot();
});

test('should create an action to load Apollo page', () => {
  expect(actions.loadApolloPage()).toMatchSnapshot();
});

test('should create a success action to load Apollo page', () => {
  expect(actions.loadApolloPageSuccess()).toMatchSnapshot();
});

test('should create a failure action to load Apollo page', () => {
  expect(actions.loadApolloPageFailure(new Error())).toMatchSnapshot();
});
