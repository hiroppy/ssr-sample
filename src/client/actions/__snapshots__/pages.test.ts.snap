// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a failure action to load Orgs page 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "LOAD_ORGS_PAGE_FAILURE",
}
`;

exports[`should create a failure action to load Top page 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "LOAD_TOP_PAGE_FAILURE",
}
`;

exports[`should create a failure action to load app process 1`] = `
Object {
  "err": [Error],
  "type": "LOAD_APP_PROCESS_FAILURE",
}
`;

exports[`should create a success action to load Orgs page 1`] = `
Object {
  "type": "LOAD_ORGS_PAGE_SUCCESS",
}
`;

exports[`should create a success action to load Top page 1`] = `
Object {
  "type": "LOAD_TOP_PAGE_SUCCESS",
}
`;

exports[`should create a success action to load app process 1`] = `
Object {
  "type": "LOAD_APP_PROCESS_SUCCESS",
}
`;

exports[`should create an action to load Orgs page 1`] = `
Object {
  "payload": Object {
    "org": "foo",
  },
  "type": "LOAD_ORGS_PAGE",
}
`;

exports[`should create an action to load Top page 1`] = `
Object {
  "type": "LOAD_TOP_PAGE",
}
`;

exports[`should create an action to load app process 1`] = `
Object {
  "type": "LOAD_APP_PROCESS",
}
`;

exports[`should create an action to reset state 1`] = `
Object {
  "type": "RESET_PAGES_STATUS",
}
`;

exports[`should create an action to stop saga 1`] = `
Object {
  "type": "STOP_SAGA",
}
`;
